pkg/pool-utils/contracts/InvariantGrowthProtocolSwapFees.sol

```diff
@@ -0,0 +1,88 @@
+        // |   increase from fees    |
+        // |-------------------------|-- original invariant * supply growth ratio (fee-less invariant)
+        // |                         |
+        // | increase from balances  |
+        // |-------------------------|-- original invariant
+        // |                         |
+        // |                         |  |------------------|-- currentSupply
+        // |                         |  |    BPT minted    |
+        // |                         |  |------------------|-- previousSupply
+        // |   original invariant    |  |  original supply |
+        // |_________________________|  |__________________|
+        //
+        // If the join is proportional, the invariant and supply will likewise increase proportionally,
+        // so the growth ratios (invariantGrowthRatio / supplyGrowthRatio) will be equal. In this case, we do not charge
+        // any protocol fees.
+        // We also charge no protocol fees in the case where `invariantGrowthRatio > supplyGrowthRatio` to avoid
+        // potential underflows, however this should not occur in normal Pool operation.
+
+        uint256 supplyGrowthRatio = currentSupply.divDown(previousSupply);
+        if ((supplyGrowthRatio >= invariantGrowthRatio) || (protocolSwapFeePercentage == 0)) return 0;
+
+        // If the join is non-proportional, the supply increase will be proportionally less than the invariant increase,
+        // since the BPT minted will be based on fewer tokens (because swap fees are not included). So the supply growth
+        // is due entirely to the balance changes, while the invariant growth also includes swap fees.
+        //
+        // To isolate the amount of increase by fees then, we multiply the original invariant by the supply growth
+        // ratio to get the "feeless invariant". The difference between the final invariant and this value is then
+        // the amount of the invariant due to fees, which we convert to a percentage by normalizing against the
+        // final invariant. This is expressed as the expression below:
+        //
+        // invariantGrowthFromFees = currentInvariant - supplyGrowthRatio * previousInvariant
+        //
+        // We then divide through by current invariant so the LHS can be identified as the fraction of the pool which
+        // is made up of accumulated swap fees.
+        //
+        // swapFeesPercentage = 1 - supplyGrowthRatio * previousInvariant / currentInvariant
+        //
+        // We then define `invariantGrowthRatio` in a similar fashion to `supplyGrowthRatio` to give the result:
+        //
+        // swapFeesPercentage = 1 - supplyGrowthRatio / invariantGrowthRatio
+        //
+        // Using this form allows us only consider the ratios of the two invariants rather than absolute values,
+        // a useful property as this is sometimes easier than calculating the full invariant twice.
```

Comment:
```suggestion
        // Using this form allows us to consider only the ratios of the two invariants, rather than their absolute
        // values: a useful property, as this is sometimes easier than calculating the full invariant twice.
```

Reply:
I don't english good.